proof/refine/RISCV64/KHeap_R.thy

```diff
@@ -4625,6 +4633,174 @@ lemma updateSchedContext_no_stack_update_corres:
+ qed
+
+lemma setReply_update_corres_Q:
+  assumes R': "\<lbrakk>reply_relation reply reply'; Q reply; Q' reply'\<rbrakk>
+               \<Longrightarrow> reply_relation (f reply) (f' (reply'::reply))"
+  and RP_same: "replyPrev_same (f' reply') reply'"
+  shows
+    "corres dc
+       (\<lambda>s. kheap s ptr = Some (kernel_object.Reply reply) \<and> Q reply)
+       (obj_at' (\<lambda>obj. obj = reply' \<and> Q' reply') ptr)
+       (set_object ptr (kernel_object.Reply (f reply)))
+       (setReply ptr (f' reply'))"
+  apply (insert R')
+  apply (rule_tac F="reply_relation reply reply' \<and> Q reply \<and> Q' reply'" in corres_req)
+   apply (drule state_relation_pspace_relation)
+   apply clarsimp
+   apply (drule (1) pspace_relation_absD)
+   apply (clarsimp simp: obj_at'_def split: if_split_asm)
+  apply (rule corres_guard_imp)
+    apply (rule setReply_replyPrev_same_corres)
+     apply fastforce
+   apply clarsimp
+  apply (clarsimp simp: obj_at'_def objBits_simps RP_same)
+  done
+
+lemmas setReply_update_corres = setReply_update_corres_Q[where Q=\<top> and Q'=\<top>, simplified]
+
+lemma updateReply_corres_Q:
+  "\<lbrakk>\<And>reply reply'. \<lbrakk>reply_relation reply reply'; Q reply; Q' reply'\<rbrakk> \<Longrightarrow> reply_relation (f reply) (f' reply');
+    \<And>reply'. replyPrev_same (f' reply') reply'\<rbrakk> \<Longrightarrow>
+   corres dc
+     (\<lambda>s. \<exists>reply. kheap s ptr = Some (kernel_object.Reply reply) \<and> Q reply)
+     (\<lambda>s'. \<exists>reply'. ko_at' reply' ptr s' \<and> Q' reply')
+     (update_reply ptr f) (updateReply ptr f')"
+  apply (clarsimp simp: update_reply_def)
+  apply (rule corres_symb_exec_l[rotated 2, OF get_object_sp])
+    apply (find_goal \<open>match conclusion in "\<lbrace>P\<rbrace> f \<exists>\<lbrace>Q\<rbrace>" for P f Q \<Rightarrow> -\<close>)
+    apply (fastforce intro: get_object_exs_valid
+                      simp: obj_at_def)
+   apply wpsimp
+   apply (clarsimp simp: obj_at_def)
+  apply (rename_tac obj)
+  apply (case_tac obj;
+         clarsimp, (solves \<open>clarsimp simp: obj_at_def is_sc_obj_def corres_underlying_def\<close>)?)
+  apply (clarsimp simp: pred_conj_def)
+  apply (rule abs_ex_lift_corres)
+  apply clarsimp
+  apply (rule corres_underlying_lift_ex2')
+  apply (rule_tac F="reply_relation reply reply' \<and> Q reply \<and> Q' reply'" in corres_req)
+   apply (drule state_relation_pspace_relation)
+   apply clarsimp
+   apply (drule (1) pspace_relation_absD)
+   apply (clarsimp simp: obj_at'_def split: if_split_asm)
+  apply (clarsimp simp: updateReply_def)
+  apply (rule corres_symb_exec_r)
+     apply (rule corres_guard_imp)
+       apply (rule_tac f=f and f'="f'" and Q=Q and Q'=Q'
+                    in setReply_update_corres_Q;
+              simp?)
+      apply (clarsimp simp: obj_at_def)
+     apply fastforce
+    apply wpsimp
+    apply (clarsimp simp: obj_at'_def)
+   apply wpsimp
+  apply wpsimp
+  apply (clarsimp simp: obj_at'_def)
```

Comment:
this pattern is a bit long-winded, any way to shorten it? I guess passing obj_at'_def to wpsimp is probably going to not go where we want, but as this is terminal, maybe a `(wpsimp | ...)+`?

Reply:
I ended up going with `apply (wpsimp | normalise_obj_at' | erule obj_at'_weakenE)+`, as an attempt to make it less likely to go down unintended paths if anything changes.